File: src/engine/arcan_led.c
/*
 * Copyright 2003-2016, Björn Ståhl
 * License: 3-Clause BSD, see COPYING file in arcan source repository.
 * Reference: http://arcan-fe.com
 * Description: LED controller interface,
 *
 * By defining LED_STANDALONE it can be built separately from the
 * main engine to allow other projects to just re-use the LED control
 * interface.
 *
 * Setting the 'ext_led' database config value to a path name pointing
 * to an existing named pipe will have the engine map that as an
 * external led controller that speaks the protocol mentioned in the
 * header.
 */
#include <stdlib.h>
#include <stdint.h>
#include <stdio.h>
#include <stdbool.h>
#include <string.h>
#include <errno.h>

#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <fcntl.h>

#ifdef USB_SUPPORT
#include <hidapi/hidapi.h>
#else
	typedef struct hid_device hid_device;
#endif

#include "arcan_led.h"
#ifndef LED_STANDALONE
#include "arcan_math.h"
#include "arcan_general.h"
#include "arcan_db.h"
#else
#define COUNT_OF(x) \
	((sizeof(x)/sizeof(0[x])) / ((size_t)(!(sizeof(x) % sizeof(0[x])))))
#endif

/* to add more device types,
 * 1. patch in the type in the controller_types enum.
 * 2. for usb-devices, add a vid/pid to type mapping in the usb_tbl
 *    or alter the init function to actually scan for it and if found,
 *    forcecontroller (get the returned index and patch type).
 * 3. modify the places marked as PROTOCOL INSERTION POINT
 */
enum controller_types {
	PACDRIVE = 1,
	ARCAN_LEDCTRL = 2
};

struct led_controller {
	enum controller_types type;
	struct led_capabilities caps;
	uint64_t devid;
	int errc;

	union {
		hid_device* handle;
		int fd;
	};

/* some need a big table, others do fine with a simple bitmask */
	uint32_t table[256];
	uint16_t ledmask;

/* for FIFOs, we may need to try open the path during writes */
	char* path;
	bool no_close;
};

struct usb_ent {
	uint16_t vid, pid;
	enum controller_types type;
	struct led_capabilities caps;
	char label[16];
};

/* one of the C fuglyness, static const struct led_capabilities paccap would
 * still yield that the initialization is not constant */
#define paccap { .variable_brightness = false, .rgb = false, .nleds = 32 }

/*
 * nleds is just max, not all of these will map due to layout variations
 * that don't necessarily (?) has a different vid/pid pair
 */
static const struct usb_ent usb_tbl[] =
{
	{.vid = 0xd209, .pid = 0x1500, .type = PACDRIVE,
		.caps = paccap, .label = "Pacdrive"},
	{.vid = 0xd209, .pid = 0x1501, .type = PACDRIVE,
		.caps = paccap, .label = "Pacdrive"},
	{.vid = 0xd209, .pid = 0x1502, .type = PACDRIVE,
		.caps = paccap, .label = "Pacdrive"},
	{.vid = 0xd209, .pid = 0x1503, .type = PACDRIVE,
		.caps = paccap, .label = "Pacdrive"},
	{.vid = 0xd209, .pid = 0x1504, .type = PACDRIVE,
		.caps = paccap, .label = "Pacdrive"},
	{.vid = 0xd209, .pid = 0x1505, .type = PACDRIVE,
		.caps = paccap, .label = "Pacdrive"},
	{.vid = 0xd209, .pid = 0x1506, .type = PACDRIVE,
		.caps = paccap, .label = "Pacdrive"},
	{.vid = 0xd209, .pid = 0x1507, .type = PACDRIVE,
		.caps = paccap, .label = "Pacdrive"},
	{.vid = 0xd209, .pid = 0x1508, .type = PACDRIVE,
		.caps = paccap, .label = "Pacdrive"}
};
#undef paccap

/*
 * not worth spending time on dynamic array management,
 * just limit to 64, have index % 64 to skip overflow checks
 */
#define MAX_LED_CONTROLLERS 64
static struct led_controller controllers[MAX_LED_CONTROLLERS] = {0};
static uint64_t ctrl_mask;
static int n_controllers = 0;

static int find_free_ind()
{
	for (size_t i = 0; i < MAX_LED_CONTROLLERS; i++)
		if ((ctrl_mask & (1 << i)) == 0)
			return i;
	return -1;
}

static struct led_controller* get_device(uint8_t devind)
{
	if (devind > MAX_LED_CONTROLLERS ||
		(ctrl_mask & (1 << devind)) == 0)
		return NULL;

	return &controllers[devind];
}

static struct led_controller* find_devid(uint64_t devid)
{
	if (devid == 0)
		return NULL;

	for (size_t i = 0; i < MAX_LED_CONTROLLERS; i++)
		if ((ctrl_mask & ((size_t)1 << i)) && controllers[i].devid == devid)
			return &controllers[i];

	return NULL;
}

static void forcecontroller(const struct usb_ent* ent)
{
	int ind = find_free_ind();
	if (-1 == ind)
		return;

/* the USB standard doesn't mandate a serial number (facepalm)
 * and hidraw doesn't really have something akin to an instance id */

	uint64_t devid = (
		(uint64_t)(ent->vid & 0xffff) << 16) | ((uint64_t)ent->pid & 0xffff);
/* other option would be to close and reopen the device as an attempt
 * to reset to a possibly safer state */
	if (find_devid(devid))
		return;

#ifdef USB_SUPPORT
	controllers[ind].handle = hid_open(ent->vid, ent->pid, NULL);
	controllers[ind].type = ent->type;
	controllers[ind].caps = ent->caps;
	if (controllers[ind].handle == NULL)
		return;

/* PROTOCOL_INSERTION_POINT */
	switch(ent->type){
	default:
	break;
	}
#else
	return;
#endif

	ctrl_mask |= 1 << ind;
	n_controllers++;
	arcan_led_added(ind, -1, ent->label);
}

int8_t arcan_led_register(int cmd_ch, int devref,
	const char* label, struct led_capabilities caps)
{
	int8_t id = find_free_ind();
	if (-1 == id)
		return -1;

	ctrl_mask |= 1 << id;
	controllers[id].devid = id;
	controllers[id].fd = cmd_ch;
	controllers[id].type = ARCAN_LEDCTRL;
	controllers[id].ledmask = 0;
	controllers[id].caps = caps;
	controllers[id].errc = 0;
	controllers[id].no_close = false;

	n_controllers++;
	arcan_led_added(id, devref, label);
	return id;
}

static int write_leddev(
	struct led_controller* dev, int ind, uint8_t* buf, size_t sz)
{
	int rv;

	if (-1 == dev->fd){
		dev->fd = open(dev->path, O_NONBLOCK | O_WRONLY);
		if (-1 == dev->fd)
			return 0;
	}

	while ((rv = write(dev->fd, buf, sz)) == -1){
		if (errno == EAGAIN)
			return 0;

		if (errno == EPIPE){
			if (!dev->no_close){
				arcan_led_remove(ind);
				return -1;
			}
			else
				break;
		}
	}

	if (rv == sz)
		return 1;

	return 0;
}

bool arcan_led_remove(uint8_t device)
{
	struct led_controller* leddev = get_device(device);
	if (!leddev)
		return false;

	switch(leddev->type){
	case ARCAN_LEDCTRL:
	break;
	default:
#ifdef USB_SUPPORT
		if (leddev->handle)
			hid_close(leddev->handle);
#endif
	break;
	}
	ctrl_mask &= ~(1 << device);
	n_controllers--;
	leddev->handle = NULL;
	arcan_led_removed(device);

	return true;
}

bool arcan_led_known(uint16_t vid, uint16_t pid)
{
	for (size_t i = 0; i < COUNT_OF(usb_tbl); i++)
		if (usb_tbl[i].vid == vid && usb_tbl[i].pid == pid)
			return true;

	return false;
}

#ifndef LED_STANDALONE
static bool register_fifo(char* path, int ind)
{
	if (ind < 0 || ind > 99)
		return false;

/* label will only be used for the event to the scripting layer, no ref */
	char buf[ sizeof("(led-fifo 10)") ];
	snprintf(buf, sizeof(buf), "(led-fifo %d)", ind);
	int ledid = arcan_led_register(-1, -1, buf,
		(struct led_capabilities){
			.nleds = 255,
			.variable_brightness = true,
			.rgb = true
		}
	);

/* delete- protect the LED so we won't close on EPIPE */
	if (-1 != ledid){
		controllers[ledid].no_close = true;
		controllers[ledid].path = path;
		return true;
	}
	return false;
}
#endif

void arcan_led_init()
{

/* register up to 10 custom LED devices that follow our FIFO protocol.
 * leddev- takes control over appl_val dynamic string ownership */
#ifndef LED_STANDALONE
	const char* appl;
	struct arcan_dbh* dbh = arcan_db_get_shared(&appl);

	char* kv = arcan_db_appl_val(dbh, appl, "ext_led");
	if (kv && register_fifo(kv, 1)){
		char work[sizeof("ext_led_1")];

		for (size_t i = 2; i < 10; i++){
			snprintf(work, sizeof(work), "ext_led_%zu", i);
			kv = arcan_db_appl_val(dbh, appl, work);
			if (!kv || !register_fifo(kv, i)){
				free(kv);
			}
		}
	}
	else
		free(kv);
#endif

	for (size_t i = 0; i < sizeof(usb_tbl) / sizeof(usb_tbl[0]); i++)
		forcecontroller(&usb_tbl[i]);
}

uint64_t arcan_led_controllers()
{
	return ctrl_mask;
}

static void ultimarc_update(uint8_t device, struct led_controller* ctrl)
{
#ifdef USB_SUPPORT
	uint8_t dbuf[] = {
		0x00, 0x00, 0xdd, (char)ctrl->ledmask, (char)(ctrl->ledmask >> 8) & 0xff};
	int val = hid_write(ctrl->handle, dbuf, sizeof(dbuf) / sizeof(dbuf[0]));
	if (-1 == val){
		ctrl->errc++;
		if (ctrl->errc > 10){
			arcan_led_remove(device);
		}
	}
#endif
}

int arcan_led_intensity(uint8_t device, int16_t led, uint8_t intensity)
{
	bool rv = false;
	struct led_controller* leddev = get_device(device);
	uint8_t di = led < 0 ? 255 : led;
	if (!leddev)
		return false;

	if (di > leddev->caps.nleds-1 && di != 255)
		return false;

	if (!leddev->caps.variable_brightness)
		intensity = intensity > 0 ? 255 : 0;

/* PROTOCOL INSERTION POINT */
	switch (controllers[device].type) {
	case ARCAN_LEDCTRL:
		if (led < 0)
			return write_leddev(leddev, device, (uint8_t[]){
				'A', '\0', 'i', intensity, 'c', '\0'}, 6);
		else
			return write_leddev(leddev, device, (uint8_t[]){
				'a', (uint8_t)led, 'i', intensity, 'c', '\0'}, 6);
	break;
	case PACDRIVE:
		if (intensity)
			leddev->ledmask |= 1 << led;
		else
			leddev->ledmask &= ~(1 << led);
		ultimarc_update(device, leddev);
	break;
	default:
		arcan_warning("Warning: arcan_led_intensity(), unknown LED / "
			"unsupported mode for device type: %i\n", controllers[device].type);
		return false;
	}

	return true;
}

int arcan_led_rgb(uint8_t device,
	int16_t led, uint8_t r, uint8_t g, uint8_t b, bool buffer)
{
	struct led_controller* leddev = get_device(device);
	if (!leddev || !leddev->caps.rgb || led > (int)leddev->caps.nleds)
		return -1;

	switch (leddev->type){
	case ARCAN_LEDCTRL:{
		int v;
		if (led < 0)
			v = write_leddev(leddev, device, (uint8_t[]){'A', '\0'}, 2);
		else
			v = write_leddev(leddev, device, (uint8_t[]){'a', (uint8_t)led}, 2);

		if (1 == v)
			return write_leddev(leddev, device, (uint8_t[]){
				'r', r, 'g', g, 'b', b, 'c', buffer ? 255 : 0}, 8);
		else
			return v;
	}
	break;
	default:
		arcan_warning("Warning: arcan_led_rgb(), unknown LED / unsupported mode"
			"	for device type: %i\n", leddev->type);
		return -1;
	}
}

struct led_capabilities arcan_led_capabilities(uint8_t device)
{
	struct led_capabilities rv = {0};
	struct led_controller* leddev = get_device(device);
	if (!leddev)
		return rv;
	return leddev->caps;
}

void arcan_led_shutdown()
{
	for (int i = 0; i < MAX_LED_CONTROLLERS && n_controllers > 0; i++)
		if ((ctrl_mask & (1 << i)) > 0){
			n_controllers--;
			switch(controllers[i].type){
			case PACDRIVE:
#ifdef USB_SUPPORT
				hid_close(controllers[i].handle);
#endif
			break;
			case ARCAN_LEDCTRL:{
				if (-1 == write(controllers[i].fd, (char[]){'o', '\0'}, 2))
					arcan_warning("arcan_led_shutdown(), error sending shutdown\n");
				close(controllers[i].fd);
				free(controllers[i].path);
				controllers[i].path = NULL;
			}
			break;
			}
		}

	ctrl_mask = 0;
}

File: src/engine/arcan_led.zig
// Pure Zig port of engine/arcan_led.c — zero C helpers.
// LED controller interface: registry, FIFO communication, protocol commands.
// No USB_SUPPORT in compositor build — no hidapi dependency.

const std = @import("std");

const c = struct {
    extern fn open(path: [*c]const u8, flags: c_int, ...) callconv(.c) c_int;
    extern fn write(fd: c_int, buf: [*c]const u8, count: usize) isize;
    extern fn close(fd: c_int) c_int;
    extern fn free(ptr: ?*anyopaque) void;

    const O_NONBLOCK: c_int = 0o4000;
    const O_WRONLY: c_int = 0o1;
    const EAGAIN: c_int = 11;
    const EPIPE: c_int = 32;
};

extern fn __errno_location() *c_int;

// Engine callbacks (defined in arcan_event.c)
extern fn arcan_led_added(devid: c_int, refdev: c_int, label: [*c]const u8) void;
extern fn arcan_led_removed(devid: c_int) void;
extern fn arcan_warning(fmt: [*c]const u8, ...) callconv(.c) void;

// Database API
const arcan_dbh = anyopaque;
extern fn arcan_db_get_shared(appl: *[*c]const u8) ?*arcan_dbh;
extern fn arcan_db_appl_val(dbh: ?*arcan_dbh, appl: [*c]const u8, key: [*c]const u8) [*c]u8;

const MAX_LED_CONTROLLERS = 64;

const ControllerType = enum(c_int) {
    pacdrive = 1,
    arcan_ledctrl = 2,
};

const led_capabilities = extern struct {
    nleds: c_int,
    variable_brightness: bool,
    rgb: bool,
};

const LedController = struct {
    ctrl_type: ControllerType = .arcan_ledctrl,
    caps: led_capabilities = .{ .nleds = 0, .variable_brightness = false, .rgb = false },
    devid: u64 = 0,
    errc: c_int = 0,
    fd: c_int = -1,
    ledmask: u16 = 0,
    path: [*c]u8 = null,
    no_close: bool = false,
};

const UsbEnt = struct {
    vid: u16,
    pid: u16,
};

const usb_tbl = [_]UsbEnt{
    .{ .vid = 0xd209, .pid = 0x1500 },
    .{ .vid = 0xd209, .pid = 0x1501 },
    .{ .vid = 0xd209, .pid = 0x1502 },
    .{ .vid = 0xd209, .pid = 0x1503 },
    .{ .vid = 0xd209, .pid = 0x1504 },
    .{ .vid = 0xd209, .pid = 0x1505 },
    .{ .vid = 0xd209, .pid = 0x1506 },
    .{ .vid = 0xd209, .pid = 0x1507 },
    .{ .vid = 0xd209, .pid = 0x1508 },
};

var controllers: [MAX_LED_CONTROLLERS]LedController = [_]LedController{.{}} ** MAX_LED_CONTROLLERS;
var ctrl_mask: u64 = 0;
var n_controllers: c_int = 0;

fn find_free_ind() c_int {
    for (0..MAX_LED_CONTROLLERS) |i| {
        if ((ctrl_mask & (@as(u64, 1) << @intCast(i))) == 0)
            return @intCast(i);
    }
    return -1;
}

fn get_device(devind: u8) ?*LedController {
    if (devind >= MAX_LED_CONTROLLERS or
        (ctrl_mask & (@as(u64, 1) << @intCast(devind))) == 0)
        return null;
    return &controllers[devind];
}

fn write_leddev(dev: *LedController, ind: c_int, buf: [*c]const u8, sz: usize) c_int {
    if (dev.fd == -1) {
        dev.fd = c.open(dev.path, c.O_NONBLOCK | c.O_WRONLY);
        if (dev.fd == -1)
            return 0;
    }

    while (true) {
        const rv = c.write(dev.fd, buf, sz);
        if (rv == -1) {
            const err = __errno_location().*;
            if (err == c.EAGAIN)
                return 0;
            if (err == c.EPIPE) {
                if (!dev.no_close) {
                    _ = arcan_led_remove(@intCast(@as(u32, @bitCast(ind))));
                    return -1;
                }
                break;
            }
            return 0;
        }
        if (rv == @as(isize, @intCast(sz)))
            return 1;
        return 0;
    }
    return 0;
}

fn register_fifo(path: [*c]u8, ind: c_int) bool {
    if (ind < 0 or ind > 99)
        return false;

    var buf: [32]u8 = undefined;
    const label = std.fmt.bufPrintZ(&buf, "(led-fifo {d})", .{ind}) catch return false;
    _ = label;

    const ledid = arcan_led_register(-1, -1, &buf, .{
        .nleds = 255,
        .variable_brightness = true,
        .rgb = true,
    });

    if (ledid != -1) {
        const uid: usize = @intCast(@as(u8, @bitCast(ledid)));
        controllers[uid].no_close = true;
        controllers[uid].path = path;
        return true;
    }
    return false;
}

export fn arcan_led_register(
    cmd_ch: c_int,
    devref: c_int,
    label: [*c]const u8,
    caps: led_capabilities,
) i8 {
    const id = find_free_ind();
    if (id == -1) return -1;

    const uid: usize = @intCast(id);
    ctrl_mask |= @as(u64, 1) << @intCast(uid);
    controllers[uid].devid = uid;
    controllers[uid].fd = cmd_ch;
    controllers[uid].ctrl_type = .arcan_ledctrl;
    controllers[uid].ledmask = 0;
    controllers[uid].caps = caps;
    controllers[uid].errc = 0;
    controllers[uid].no_close = false;

    n_controllers += 1;
    arcan_led_added(id, devref, label);
    return @intCast(id);
}

export fn arcan_led_remove(device: u8) bool {
    const leddev = get_device(device) orelse return false;
    _ = leddev;

    ctrl_mask &= ~(@as(u64, 1) << @intCast(device));
    n_controllers -= 1;
    arcan_led_removed(@intCast(device));
    return true;
}

export fn arcan_led_known(vid: u16, pid: u16) bool {
    for (usb_tbl) |ent| {
        if (ent.vid == vid and ent.pid == pid)
            return true;
    }
    return false;
}

export fn arcan_led_init() void {
    var appl: [*c]const u8 = null;
    const dbh = arcan_db_get_shared(&appl);

    var kv = arcan_db_appl_val(dbh, appl, "ext_led");
    if (kv != null and register_fifo(kv, 1)) {
        for (2..10) |i| {
            var work: [16]u8 = undefined;
            const key = std.fmt.bufPrintZ(&work, "ext_led_{d}", .{i}) catch break;
            _ = key;
            kv = arcan_db_appl_val(dbh, appl, &work);
            if (kv == null or !register_fifo(kv, @intCast(i))) {
                c.free(kv);
            }
        }
    } else {
        c.free(kv);
    }
}

export fn arcan_led_controllers() u64 {
    return ctrl_mask;
}

export fn arcan_led_intensity(device: u8, led: i16, intensity_arg: u8) c_int {
    const leddev = get_device(device) orelse return 0;
    const di: u16 = if (led < 0) 255 else @intCast(led);

    if (leddev.caps.nleds > 0 and di != 255 and di > @as(u16, @intCast(leddev.caps.nleds - 1)))
        return 0;

    var intensity = intensity_arg;
    if (!leddev.caps.variable_brightness)
        intensity = if (intensity > 0) 255 else 0;

    switch (controllers[device].ctrl_type) {
        .arcan_ledctrl => {
            if (led < 0)
                return write_leddev(leddev, @intCast(device), &[_]u8{ 'A', 0, 'i', intensity, 'c', 0 }, 6)
            else
                return write_leddev(leddev, @intCast(device), &[_]u8{ 'a', @intCast(led), 'i', intensity, 'c', 0 }, 6);
        },
        .pacdrive => {
            return 0;
        },
    }
}

export fn arcan_led_rgb(device: u8, led: i16, r: u8, g: u8, b: u8, buffer: bool) c_int {
    const leddev = get_device(device) orelse return -1;
    if (!leddev.caps.rgb or led > @as(i16, @intCast(leddev.caps.nleds)))
        return -1;

    switch (leddev.ctrl_type) {
        .arcan_ledctrl => {
            const v = if (led < 0)
                write_leddev(leddev, @intCast(device), &[_]u8{ 'A', 0 }, 2)
            else
                write_leddev(leddev, @intCast(device), &[_]u8{ 'a', @intCast(led) }, 2);

            if (v == 1)
                return write_leddev(leddev, @intCast(device), &[_]u8{ 'r', r, 'g', g, 'b', b, 'c', if (buffer) 255 else 0 }, 8)
            else
                return v;
        },
        .pacdrive => {
            return -1;
        },
    }
}

export fn arcan_led_capabilities(device: u8) led_capabilities {
    const leddev = get_device(device) orelse
        return .{ .nleds = 0, .variable_brightness = false, .rgb = false };
    return leddev.caps;
}

export fn arcan_led_shutdown() void {
    var i: usize = 0;
    while (i < MAX_LED_CONTROLLERS and n_controllers > 0) : (i += 1) {
        if ((ctrl_mask & (@as(u64, 1) << @intCast(i))) != 0) {
            n_controllers -= 1;
            switch (controllers[i].ctrl_type) {
                .pacdrive => {},
                .arcan_ledctrl => {
                    _ = c.write(controllers[i].fd, &[_]u8{ 'o', 0 }, 2);
                    _ = c.close(controllers[i].fd);
                    c.free(controllers[i].path);
                    controllers[i].path = null;
                },
            }
        }
    }
    ctrl_mask = 0;
}
